### PR TITLE
Added `pause` to core.atomic

### PR DESCRIPTION
**Pull request by TurkeyMan**
So that other architectures can add their special sauce.

### CODE REVIEW

File: src/core/atomic.d [hunk 1]
```diff
@@ -466,7 +466,7 @@ in (atomicPtrIsProperlyAligned(here), "Argument `here` is not properly aligned")
 *  true if the store occurred, false if not.
 */
 bool casWeak(MemoryOrder succ = MemoryOrder.seq,MemoryOrder fail = MemoryOrder.seq,T,V)(T* here, T* ifThis, V writeThis) pure nothrow @nogc @trusted
-if (!is(T == shared S, S) && !is(V == shared U, U))
+    if (!is(T == shared S, S) && !is(V == shared U, U))
```
Comment (wilzbach) [inline src/core/atomic.d]: Before your change this was the official DStyle (dlang.org/dstyle).

Comment (TurkeyMan) [reply src/core/atomic.d]: It was inconsistent with the rest of the file... I would always favour internal consistency :/

**Comment by CyberShadow** [comment]
Maybe refactor the one in `core.internal.spinlock` to use your implementation?

**Comment by TurkeyMan** [comment]
Those spinlock backoff counters look sus; they're *very* small numbers.
With backoff set to 4/16 cycles, it's not really a spinlock at all! It's just a sleep-lock.
NT kernel uses ~4000 cycles before backoff... and it probably shouldn't be called a "spin lock" if it backs off.

**Comment by CyberShadow** [comment]
> Those spinlock backoff counters look sus; they're _very_ small numbers.

Yes, [I ran into issues with that too](https://github.com/dlang/druntime/pull/2817#discussion_r331786943).

**Comment by TurkeyMan** [comment]
Yeah, it's not a spin-lock... and it's also really bad >_<

**Comment by TurkeyMan** [comment]
Is this controversial?

File: changelog/atomic_cas.dd [hunk 2]
```diff
@@ -1,4 +1,4 @@
-Added overloads for `core.atomic.cas` which don't discard their result.
+Added overloads for `core.atomic : cas` which don't discard their result
```
Comment (MoonlightSentinel) [inline changelog/atomic_cas.dd]: Unrelated style change
Ditto for all other changelogs and template constraints

File: src/core/atomic.d [hunk 3]
```diff
@@ -519,11 +519,20 @@ in (atomicPtrIsProperlyAligned(here), "Argument `here` is not properly aligned")
  * that all loads and stores before a call to this function are executed before any
  * loads and stores after the call.
  */
-void atomicFence(MemoryOrder order = MemoryOrder.seq)() nothrow @nogc @safe
+void atomicFence(MemoryOrder order = MemoryOrder.seq)() pure nothrow @nogc @safe
```
Comment (MoonlightSentinel) [inline src/core/atomic.d]: Unrelated to pause(), please move this into another PR
Ditto for all other `pure` additions

File: src/core/internal/spinlock.d [hunk 4]
```diff
@@ -52,7 +52,7 @@ shared struct SpinLock
     {
         import core.time;
         if (k < pauseThresh)
-            return pause();
+            return core.atomic.pause();
```
Comment (MoonlightSentinel) [inline src/core/internal/spinlock.d]: Correct if im wrong but this is more than a simple refactoring as the new inline assembly differs:
Old:
```
rep
nop
```
New:
```
naked
rep
nop
ret
```

File: src/core/internal/atomic.d [hunk 5]
```diff
@@ -645,6 +645,33 @@ void atomicFence(MemoryOrder order = MemoryOrder.seq)() nothrow @nogc @safe
...
+    version (D_InlineAsm_X86)
+    {
+        asm pure nothrow @nogc @trusted
+        {
+            naked;
+            rep; nop;
+            ret;
+        }
+    }
+    else version (D_InlineAsm_X86_64)
```
Comment (MoonlightSentinel) [inline src/core/internal/atomic.d]: Keep the old enum/version declaration instead of duplicating the assembly? Or should there be different implementations according to your comment regarding the `pause` opcode?

**Comment by MoonlightSentinel** [comment]
Damn ...

File: src/core/atomic.d (hunk 1 above)
Comment (wilzbach) [reply src/core/atomic.d]: We have a style guide for a reason!!
